lusidtools/cocoon/cocoon.py

```diff
@@ -1673,6 +1671,38 @@ def load_from_data_frame(
             property_columns=property_columns,
         )
 
+    # If the transaction contains subholding keys which aren't defined in the portfolio. We first create the
+    # properties that don't already exist, then we make the properties sub-holding keys in the portfolios.
+    if file_type == "transaction" and sub_holding_keys is not None and sub_holding_keys != []:
+        # if the SHK key is written in {domain}/{scope}/{code} form we extract the code since when we add it to
+        # the properties there will be issues due to different formats. Also, there are issues with the
+        # create_missing_property_definitions_from_file function as it extracts data from columns using the
+        # sub-holding keys.
+        sub_holding_keys_codes = [key if '/' not in key else key.split('/')[2] for key in sub_holding_keys]
+
+        # Check for and create missing property definitions for the sub-holding-keys
+        data_frame = cocoon.properties.create_missing_property_definitions_from_file(
+            api_factory=api_factory,
+            properties_scope=properties_scope,
+            domain="Transaction",
+            data_frame=data_frame,
+            property_columns=[{"source": key} for key in sub_holding_keys_codes],
+        )
+
+        # Add subholding keys to the portfolios we are going to apply the transactions to
+        for code in set(data_frame[mapping_required['code']]):
+            api_factory.build(
```

Comment:
For this loop, you are creating a new TransactionPortfoliosApi every time, it just needs to be created once before the loop and then accessed inside of it